Firestore/core/src/firebase/firestore/local/serializer.cc

```diff
@@ -0,0 +1,155 @@
+      writer->WriteNestedMessage([&](Writer* writer) {
+        const model::Document& doc =
+            static_cast<const model::Document&>(maybe_doc);
+        rpc_serializer_.EncodeDocument(writer, doc.key(),
+                                       doc.data().object_value());
+      });
+      return;
+
+    case model::MaybeDocument::Type::NoDocument:
+      // TODO(rsgowman)
+      abort();
+
+    case model::MaybeDocument::Type::Unknown:
+      // TODO(rsgowman)
+      abort();
+  }
+
+  UNREACHABLE();
+}
+
+std::unique_ptr<model::MaybeDocument> Serializer::DecodeMaybeDocument(
+    Reader* reader) const {
+  if (!reader->status().ok()) return nullptr;
+
+  // Initialize MaybeDocument fields to their default values. (Due to the
+  // 'oneof' in MaybeDocument, only one of 'no_document' or 'document' should
+  // ever be set.)
+  std::unique_ptr<model::NoDocument> no_document;
+  std::unique_ptr<model::Document> document;
+
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return nullptr;
+
+    // Ensure the tag matches the wire type
+    switch (tag.field_number) {
+      case firestore_client_MaybeDocument_document_tag:
+      case firestore_client_MaybeDocument_no_document_tag:
+        if (tag.wire_type != PB_WT_STRING) {
+          reader->set_status(
+              Status(FirestoreErrorCode::DataLoss,
+                     "Input proto bytes cannot be parsed (mismatch between "
+                     "the wiretype and the field number (tag))"));
+        }
+        break;
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these. No
+        // action required here, though we'll need to skip the relevant bytes
+        // below.
+        break;
+    }
+
+    if (!reader->status().ok()) return nullptr;
+
+    switch (tag.field_number) {
+      case firestore_client_MaybeDocument_document_tag:
+        // 'no_document' and 'document' are part of a oneof. The proto docs
+        // claim that if both are set on the wire, the last one wins.
+        no_document = nullptr;
+
+        // TODO(rsgowman): If multiple '_document' values are found, we should
+        // merge them (rather than using the last one.)
+        document = reader->ReadNestedMessage<std::unique_ptr<model::Document>>(
+            [&](Reader* reader) -> std::unique_ptr<model::Document> {
+              return rpc_serializer_.DecodeDocument(reader);
+            });
+        break;
+
+      case firestore_client_MaybeDocument_no_document_tag:
+        // 'no_document' and 'document' are part of a oneof. The proto docs
+        // claim that if both are set on the wire, the last one wins.
+        document = nullptr;
+
+        // TODO(rsgowman): Parse the no_document field.
+        abort();
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these.
+        reader->SkipField(tag);
```

Comment:
I feel the logic is tore apart by the two switches. I'd move the logic here above, i.e. in previous default, just do
```
reader->SkipField(tag);
continue;
```
and put unreachable here.

Reply:
My reasoning behind this was to have the first switch handle the type verification. (e.g. ensure string fields contains strings, int fields contain ints.) And then the second switch can just focus on parsing the contents of the field. I like this separation but am willing to be out-voted. Also note that while it's not entirely obvious here (due to there only being 2 fields) in other messages there are many more. In these cases, the type verification reduces to:

```
switch(field_number) {
  case tag1:
  case tag3:
  case tag5:
    assert(type == string)
    break;
  case 2:
  case 4:
  case 6:
    assert(type == varint)
    break;
  default:
    //nothing!
    break;
}
```

There's typically only the 3 "real" blocks (assuming you count the empty default block). So the type verification switch statement is usually pretty simple, whereas the parsing one is usually more complex.

Although I haven't gotten to it yet, I suspect the first switch statement could be refactored to a single helper method call: `VerifyTypeForMaybeDocumentField(...)` or similar which might address your concerns somewhat? At the very least, it should make it more obvious what's going on.

Regardless of what we decide, the pattern here is used throughout the serializers, so I'll apply the final resolution (if any) everywhere in a followup.

Options:
1) Two switches. (i.e. as currently written)
2) Refactor into `VerifyTypesForMaybeDocumentField`. Two switches still remain, but not in the same method. The parsing itself might be refactored too to make this method even simpler.
3) Merge two switches (verify types + parse in the same case block.)

Reply:
Or were you referring only to the 'default' case?

Reply:
I'm OK with the other cases's separation. For the `default` case, the logic is simple and might be combined. It is in the same dogma of type
```
func {
  if (foo) {
     bar;
     return;
  }
  // now in below code, we assume !foo.
  ...
}
```

Reply:
Basically, an early exit. I'd like to avoid that to keep all the reading together in the second switch. It should also allow some potential future refactoring, i.e.:

```
DecodeFoo() {
  ...
  while (field = GetNextField()) {
    EnsureTagAndTypeCorrect(field.tag, field.type);  // switch #1 hidden in here
    DecodeField(field);  // switch #2 hidden in here
  }
  ...
}
```

(We could still combine the two default cases and still have the above work if the first function returned some sort of early exit indicator.)

Reply:
I was about to comment separately about the two switch thing, so I'll cast my vote for just a single switch statement.

I think this could be made less onerous by adding a method to the reader that allows you to call it like so:

```c++
reader->RequireWireType(PT_WT_STRING, tag);
```

You can then just make this the first line of each case.

I favor this approach because which type applies and the code to handle it are not independent. It would be too easy to change code here to use e.g. an int and then forget to break out the same case above.

You could also have `ReadNestedMessage` take the parent tag as a parameter and have it validate that the parent has the correct wire type (which AIUI will always be PB_WT_STRING for nested messages).

Reply:
Alright; I'll author a followup that merges the double-switch pattern wherever it occurs.